src/lib/configs/project/spell.ts

```diff
@@ -0,0 +1,294 @@
+        minimum: 0,
+        maximum: U32_MAX,
+      },
+    },
+    required: [],
+  },
+} as const;
+
+export const overridableSpellProperties = {
+  ...spellProperties,
+  aquaFilePath: {
+    ...spellProperties.aquaFilePath,
+    nullable: true,
+  },
+  function: {
+    ...spellProperties.function,
+    nullable: true,
+  },
+} as const;
+
+const configSchemaV0: JSONSchemaType<ConfigV0> = {
+  type: "object",
+  $id: `${TOP_LEVEL_SCHEMA_ID}/${SPELL_CONFIG_FILE_NAME}`,
+  title: SPELL_CONFIG_FILE_NAME,
+  description: `Defines a spell. You can use \`fluence spell new\` command to generate a template for new spell`,
+  properties: spellProperties,
+  required: ["version", "function", "aquaFilePath"],
+};
+
+const migrations: Migrations<Config> = [];
+type Config = ConfigV0;
+type LatestConfig = ConfigV0;
+export type SpellConfig = InitializedConfig<LatestConfig>;
+export type SpellConfigReadonly = InitializedReadonlyConfig<LatestConfig>;
+
+const getDateSec = (date: Date) => Math.round(date.getTime() / 1000);
+
+export const resolveStartSec = ({ clock }: LatestConfig): number => {
+  if (clock === undefined) {
+    return 0;
+  }
+
+  if (clock.startDelaySec !== undefined) {
+    return getDateSec(new Date()) + clock.startDelaySec;
+  }
+
+  return clock.startTimestamp === undefined
+    ? 1
+    : getDateSec(new Date(clock.startTimestamp));
+};
+
+export const resolveEndSec = ({ clock }: LatestConfig): number => {
+  if (clock === undefined) {
+    return 0;
+  }
+
+  if (clock.endDelaySec !== undefined) {
+    return getDateSec(new Date()) + clock.endDelaySec;
+  }
+
+  return clock.endTimestamp === undefined
+    ? 0
+    : getDateSec(new Date(clock.endTimestamp));
+};
+
+const validate: ConfigValidateFunction<LatestConfig> = async (
+  config,
+  configPath
+) =>
+  validateBatch(
+    config.clock?.startTimestamp !== undefined &&
+      config.clock?.startDelaySec !== undefined
+      ? `You can't specify both 'startTimestamp' and 'startDelaySec' properties`
+      : true,
+
+    config.clock?.endTimestamp !== undefined &&
+      config.clock?.endDelaySec !== undefined
+      ? `You can't specify both 'endTimestamp' and 'endDelaySec' properties`
+      : true,
+
+    (() => {
+      let endSec;
+
+      try {
+        endSec = resolveEndSec(config);
+      } catch {
+        return "Invalid `endTimestamp` value. It must be an ISO timestamp";
+      }
+
+      try {
+        const startSec = resolveStartSec(config);
+
+        if (startSec === 0) {
+          return true;
+        }
+
+        if (endSec !== 0 && endSec < startSec) {
+          return `Start time must be earlier than end time. Got: start time ${startSec}, end time ${endSec}`;
+        }
+
+        const currentSec = getDateSec(new Date());
+
+        if (endSec !== 0 && endSec < currentSec) {
+          return `End time must be later than current time. Got: current time ${currentSec}, end time ${endSec}`;
+        }
+
+        if (endSec - startSec > MAX_PERIOD_SEC) {
```

Comment:
The period to which this restriction applies is `periodSec` field, not `endSec - startSec`.

Reply:
Oh clearly. Thanks!